### PR TITLE
SYMMLQ improvements

### PR DESCRIPTION
**Pull request by amontoison**
I modified SYMMLQ to be able to compute `x_lq` when `x_cg` doesn't exist.
`ζ = ζbar * c` is equal to `Inf` if `ζbar = Inf`  / `γbar = 0`.
The `η` variable has been introduce to solve this problem.

### CODE REVIEW

File: src/symmlq.jl [hunk 1]
```diff
@@ -82,7 +85,7 @@ function symmlq(A :: AbstractLinearOperator, b :: AbstractVector{T};
   ANorm = 0.0
   Acond = 0.0
   rNorm = β₁
-  rcgNorm = β₁
+  rcgNorm = β * abs(ζbar)
```
Comment (dpo) [inline src/symmlq.jl]: Can't we still initialize this to beta1 with a comment that it is equal to beta * abs(zetabar)?

File: src/symmlq.jl [hunk 2]
```diff
@@ -172,10 +173,11 @@ function symmlq(A :: AbstractLinearOperator, b :: AbstractVector{T};
     γbar = δbar * s - α * c
     ϵ = β * s
     δbar = -β * c
-    ζbar = -(ϵold * ζold + δ * ζ)/γbar
+    η = -ϵold * ζold - δ * ζ
+    ζbar = η / γbar
 
     rNorm = sqrt(γ * γ * ζ * ζ + ϵold * ϵold * ζold * ζold)
-    rcgNorm = abs(rcgNorm*s*cold/c)
+    rcgNorm = β * abs(s * ζ - c * ζbar)
```
Comment (dpo) [inline src/symmlq.jl]: Could you provide more information on these two changes?

Comment (dpo) [reply src/symmlq.jl]: Should we check if gammabar = 0 ?

**Comment by amontoison** [comment]
I rebase this pull request.
Contrary to `BiLQ` or `USYMLQ`, I don't check that `γbar ≠ 0` when I compute `ζbar`. We need it to compute `CG` residual norms and error norms bounds that are returned in a `SymmlqStats` at the end.

**Comment by dpo** [comment]
I'm not thrilled about having Infs in the code when gammabar = 0. I think we should test like in the other solvers?!

**Comment by amontoison** [comment]
My problem is what should we do with `rcgNorm` and `errorscg` if `gammabar = 0`?
A solution could be to not update them, but the drawback is that we loose the link between `iter` and `rcgNorm[iter]`. We don't anymore at which iteration this error norm or residual norm correspond.

**Comment by dpo** [comment]
How do you do it in BiLQ? We could insert a `missing` value for the CG residual and error, which would avoid using Infs and NaNs.

**Comment by amontoison** [comment]
In BiLQ and USYMLQ, residual norms are computed only if CG point exists (δbarₖ ≠ 0). With those methods, a SimpleStats is returned at the end. So it's not a problem if rcgNorms are not computed because we don't need to store them.
With SYMMLQ, more elaborated statistics are returned. They include residuals and errors norms of CG points. Using missing value if a CG point doesn't exist is a good idea.

**Comment by dpo** [comment]
> In BiLQ and USYMLQ, residual norms are computed only if CG point exists (δbarₖ ≠ 0). With those methods, a SimpleStats is returned at the end. So it's not a problem if rcgNorms are not computed because we don't need to store them.

BiLQ and USYMLQ could (should) also return CG residuals, even if they don't form the CG iterate until the end.

**Comment by amontoison** [comment]
I think we could generalize `SymmlqStats` to `LQStats` in an other pull request for all LQ methods (LSLQ, LNLQ, SYMMLQ, BiLQ, USYMLQ).

**Comment by dpo** [comment]
sounds good.

**Comment by dpo** [comment]
Thank you.

**Pull request merged**